python-package/lightgbm/engine.py

```diff
@@ -188,8 +188,17 @@ def train(
         params=params,
         default_value=None,
     )
-    if params["early_stopping_round"] is None:
-        params.pop("early_stopping_round")
+    if "early_stopping_round" in params:
+        if params["early_stopping_round"] is None:
+            params.pop("early_stopping_round")
+        # the check below happens if the callback is instantiated, but if the user
+        # passes a non-numeric value `params.get("early_stopping_round", 0) > 0` below
+        # will fail prior to the callback instantiation, so a TypeError should be raised
+        # here as well
+        elif not isinstance(params["early_stopping_round"], int):
+            raise TypeError(
+                f"stopping_rounds should be an integer. Got {type(params['early_stopping_round'])}")
```

Comment:
Can you please remove the `if "early_stopping_round" in params` condition? That specific key is guaranteed to exist in `params` at this point because of the `_choose_param_value()` call immediately above this.

For the rest of this... I'm nervous about this amount of complexity being copied in so many places. To be clear, what you've added here is totally correct and valuable (replaces a cryptic error like "`ValueError: invalid literal for int() with base 10`" with an informative one), but I think there is a way to do it with less code duplication.

Here's my proposal... what if we add an internal function in `callback.py` like this:

```python
def _should_enable_early_stopping(stopping_rounds: Any) -> bool:
    if not isinstance(stopping_rounds, int):
        raise TypeError(f"early_stopping_round should be an integer. Got {type(stopping_rounds)}")
    return stopping_rounds > 0
```

And then call it in all 3 places? Like this in `cv()` and `train()`:

```python
if callback._should_enable_early_stopping(params.get("early_stopping_round", 0)):
    ...
```

And this in the callback

```python
self.enabled = _should_enable_early_stopping(stopping_rounds)
```

That would guarantee that you get the exact same error message in `train()`, `cv()`, or instantiating the `early_stopping()` callback directly. It'd also consolidate the logic in one place, reducing of bugs exactly like the one that led to this (#6401).

What do you think?

Reply:
I like this a lot better than what I coded.  It makes sense to me.  So really we are just moving the check as to whether to enable early stopping to ONE place and also type checking in the same function with an informative error.  I'll make those changes.  Thanks!

Reply:
🎉 thanks for considering it.